### PR TITLE
Chapter 11

### PR DESCRIPTION
**Pull request by heimrih**
Chapter 11 - Activation

### CODE REVIEW

**Comment by heimrih** [comment]
ready

File: app/controllers/account_activations_controller.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
+class AccountActivationsController < ApplicationController
+  def edit
+    user = User.find_by email: params[:email]
+    if user && !user.activated? && user.authenticated?(:activation, params[:id])
+      user.activate
+      user.update_attribute :activated, true
+      user.update_attribute :activated_at, Time.zone.now
```
Comment (saobv-2223) [inline app/controllers/account_activations_controller.rb]: In activate action, you wrote updating for activated and activated_at columns, therefore you don't need to write line 6 & 7 here 
```suggestion
```

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -4,9 +4,16 @@ def new; end
   def create
     user = User.find_by email: params[:session][:email].downcase
     if user&.authenticate params[:session][:password]
-      log_in user
-      params[:session][:remember_me] == "1" ? remember(user) : forget(user)
-      redirect_back_or user
+      if user.activated?
+        log_in user
+        params[:session][:remember_me] == '1' ? remember(user) : forget(user)
```
Comment (saobv-2223) [inline app/controllers/sessions_controller.rb]: move 1 to config and call it here

File: app/mailers/user_mailer.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
+class UserMailer < ApplicationMailer
+
+  # Subject can be set in your I18n file at config/locales/en.yml
+  # with the following lookup:
+  #
+  #   en.user_mailer.account_activation.subject
+  #
```
Comment (saobv-2223) [inline app/mailers/user_mailer.rb]: remove all the comment lines, please

File: app/models/user.rb [hunk 4]
```diff
@@ -1,5 +1,7 @@
 class User < ApplicationRecord
-  attr_accessor :remember_token
+  attr_accessor :remember_token, :activation_token
+  before_save :downcase_email
+  before_create :create_activation_digest
```
Comment (saobv-2223) [inline app/models/user.rb]:
```suggestion
  attr_accessor :remember_token, :activation_token
  
  before_save :downcase_email
  before_create :create_activation_digest
  
```

File: app/models/user.rb [hunk 5]
```diff
@@ -36,12 +38,33 @@ def remember
...
+    BCrypt::Password.new(digest).is_password? token
+end
 
   def forget
     update_attribute :remember_digest, nil
   end
+
+  def activate
+    update_attributes :activated, true
+    update_attributes :activated_at, Time.zone.now
```
Comment (saobv-2223) [inline app/models/user.rb]:
```suggestion
    update activated: true, activated_at: Time.zone.now
```

File: app/views/user_mailer/account_activation.html.erb [hunk 6]
```diff
@@ -0,0 +1,8 @@
+<h1><%= t "sampleapp" %></h1>
+<p>Hi <%= @user.name %>,</p>
+<p>
+<%= t "sample.welcome" %>
+<%= t "sample.activation" %>
```
Comment (saobv-2223) [inline app/views/user_mailer/account_activation.html.erb]: indenting 2 spaces

**Comment by heimrih** [comment]
ready

File: app/controllers/sessions_controller.rb [hunk 7]
```diff
@@ -4,9 +4,16 @@ def new; end
   def create
     user = User.find_by email: params[:session][:email].downcase
     if user&.authenticate params[:session][:password]
-      log_in user
-      params[:session][:remember_me] == "1" ? remember(user) : forget(user)
-      redirect_back_or user
+      if user.activated?
+        log_in user
+        params[:session][:remember_me] == Settings.params.one ? remember(user) : forget(user)
```
Comment (saobv-2223) [inline app/controllers/sessions_controller.rb]:
```suggestion
        params[:session][:remember_me].eql? Settings.params.one ? remember(user) : forget(user)
```

File: app/models/user.rb [hunk 8]
```diff
@@ -36,12 +39,32 @@ def remember
...
   end
 
-  def authenticated? remember_token
-    return false unless remember_digest
-    BCrypt::Password.new(remember_digest).is_password?remember_token
-  end
+  def authenticated? attribute, token
+    digest = send "#{attribute}_digest"
+    return false if digest.nil?
+    BCrypt::Password.new(digest).is_password? token
```
Comment (saobv-2223) [inline app/models/user.rb]:
```suggestion
    return false if digest.nil?
    
    BCrypt::Password.new(digest).is_password? token
```

**Comment by heimrih** [comment]
ready

File: app/controllers/sessions_controller.rb [hunk 9]
```diff
@@ -4,9 +4,16 @@ def new; end
...
-      params[:session][:remember_me] == "1" ? remember(user) : forget(user)
-      redirect_back_or user
+      if user.activated?
+        log_in user
+        params[:session][:remember_me].eql? Settings.params.one ? remember(user) : forget(user)
+        redirect_back_or user
+      else
+        message = t "user.activation.message.notactivated"
+        message += t "user.activation.message.check"
+        flash[:warning] = message
```
Comment (manhnt-0960) [inline app/controllers/sessions_controller.rb]: you should write one line

File: app/helpers/account_activations_helper.rb [hunk 10]
```diff
@@ -0,0 +1,2 @@
+module AccountActivationsHelper
```
Comment (manhnt-0960) [inline app/helpers/account_activations_helper.rb]: delete this file because u don't use

File: app/models/user.rb [hunk 11]
```diff
@@ -36,12 +39,33 @@ def remember
     update_attribute :remember_digest, User.digest(remember_token)
   end
 
-  def authenticated? remember_token
-    return false unless remember_digest
-    BCrypt::Password.new(remember_digest).is_password?remember_token
-  end
+  def authenticated? attribute, token
+    digest = send "#{attribute}_digest"
+    return false if digest.nil?
```
Comment (manhnt-0960) [inline app/models/user.rb]:
```suggestion
    return false unless digest
```

File: app/views/user_mailer/account_activation.html.erb [hunk 12]
```diff
@@ -0,0 +1,8 @@
+<h1><%= t "sampleapp" %></h1>
+<p><%= @user.name %></p>
+<p>
+  <%= t "sample.welcome" %>
+  <%= t "sample.activation" %>
+</p>
+<%= link_to t("user.activation.activate"),
+edit_account_activation_url(id: @user.activation_token, email: @user.email) %>
```
Comment (manhnt-0960) [inline app/views/user_mailer/account_activation.html.erb]: indent

File: app/views/user_mailer/account_activation.text.erb [hunk 13]
```diff
@@ -0,0 +1,6 @@
+<%= t "sampleapp" %>
+<%= @user.name %>
+<%= t "sample.welcome" %>
+<%= t "sample.activation" %>
+<%= link_to t("user.activation.activate"),
+edit_account_activation_url(id: @user.activation_token, email: @user.email) %>
```
Comment (manhnt-0960) [inline app/views/user_mailer/account_activation.text.erb]: indent

**Comment by heimrih** [comment]
ready

**Comment by manhnt-0960** [comment]
commented
fix 2 commits

**Comment by heimrih** [comment]
ready

File: app/mailers/user_mailer.rb [hunk 14]
```diff
@@ -0,0 +1,12 @@
...
+  def account_activation user
+    @user = user
+    mail to: user.email, subject: t("user.activation.activation")
+  end
+
+  def password_reset
+    @greeting = "Hi"
+
+    mail to: "to@example.org"
+  end
```
Comment (saobv-2223) [inline app/mailers/user_mailer.rb]: remove function that not used

File: app/models/user.rb [hunk 15]
```diff
@@ -1,5 +1,8 @@
 class User < ApplicationRecord
-  attr_accessor :remember_token
+  attr_accessor :remember_token, :activation_token
+
+  before_save :downcase_email
+  before_create :create_activation_digest
   USERS_PARAMS = %i(name email password password_confirmation).freeze
```
Comment (saobv-2223) [inline app/models/user.rb]: image

File: app/models/user.rb [hunk 16]
```diff
@@ -36,12 +39,33 @@ def remember
     update_attribute :remember_digest, User.digest(remember_token)
   end
 
-  def authenticated? remember_token
-    return false unless remember_digest
-    BCrypt::Password.new(remember_digest).is_password?remember_token
-  end
+  def authenticated? attribute, token
+    digest = send "#{attribute}_digest"
```
Comment (saobv-2223) [inline app/models/user.rb:42]: Can I ask you a question, what is send method ?

File: app/views/user_mailer/account_activation.html.erb [hunk 17]
```diff
@@ -0,0 +1,7 @@
+<h1><%= t "sampleapp" %></h1>
+<p><%= @user.name %></p>
+<p>
+  <%= t "sample.welcome" %>
+  <%= t "sample.activation" %>
+</p>
+  <%= link_to t("user.activation.activate"), edit_account_activation_url(id: @user.activation_token, email: @user.email) %>
```
Comment (saobv-2223) [inline app/views/user_mailer/account_activation.html.erb]:
```suggestion
<%= link_to t("user.activation.activate"), edit_account_activation_url(id: @user.activation_token, email: @user.email) %>
```

File: app/views/user_mailer/account_activation.text.erb [hunk 18]
```diff
@@ -0,0 +1,5 @@
+<%= t "sampleapp" %>
+<%= @user.name %>
+  <%= t "sample.welcome" %>
+  <%= t "sample.activation" %>
+  <%= link_to t("user.activation.activate"), edit_account_activation_url(id: @user.activation_token, email: @user.email) %>
```
Comment (saobv-2223) [inline app/views/user_mailer/account_activation.text.erb]:
```suggestion
<%= t "sample.welcome" %>
<%= t "sample.activation" %>
<%= link_to t("user.activation.activate"), edit_account_activation_url(id: @user.activation_token, email: @user.email) %>
```

File: app/views/user_mailer/password_reset.html.erb [hunk 19]
```diff
@@ -0,0 +1,5 @@
+<h1>User#password_reset</h1>
+
```
Comment (saobv-2223) [inline app/views/user_mailer/password_reset.html.erb]: remove file that not used

File: app/views/user_mailer/password_reset.text.erb [hunk 20]
```diff
@@ -0,0 +1,3 @@
+User#password_reset
```
Comment (saobv-2223) [inline app/views/user_mailer/password_reset.text.erb]: remove file that not used

File: config/settings.yml [hunk 21]
```diff
@@ -21,3 +21,6 @@ validations:
 gravartar:
   size_default: 80
   size_list: 50
+
+params:
+  one: "1"
```
Comment (saobv-2223) [inline config/settings.yml]:
```suggestion
  one: 1
```

File: app/models/user.rb (hunk 16 above)
Comment (heimrih) [reply app/models/user.rb:42]: Send method allow us to call a method

**Comment by heimrih** [comment]
ready

File: app/models/user.rb [hunk 22]
```diff
@@ -16,6 +17,8 @@ class User < ApplicationRecord
   has_secure_password
 
   before_save { email.downcase! }
+  before_save :downcase_email
```
Comment (saobv-2223) [inline app/models/user.rb]: remove before_save at line 20

**Comment by heimrih** [comment]
ready

**Comment by heimrih** [comment]
ready

File: app/controllers/sessions_controller.rb [hunk 23]
```diff
@@ -4,9 +4,15 @@ def new; end
   def create
     user = User.find_by email: params[:session][:email].downcase
     if user&.authenticate params[:session][:password]
-      log_in user
-      params[:session][:remember_me] == "1" ? remember(user) : forget(user)
-      redirect_back_or user
+      if user.activated?
+        log_in user
+        params[:session][:remember_me].eql? Settings.params.one ? remember(user) : forget(user)
```
Comment (chungpv-1008) [inline app/controllers/sessions_controller.rb]:
```suggestion
        params[:session][:remember_me].eql?(Settings.params.one) ? remember(user) : forget(user)
```

File: app/controllers/sessions_controller.rb [hunk 24]
```diff
@@ -4,9 +4,15 @@ def new; end
...
+        redirect_back_or user
+      else
+        message = t "user.activation.message"
+        flash[:warning] = message
```
Comment (chungpv-1008) [inline app/controllers/sessions_controller.rb]:
```suggestion
        flash[:warning] = t "user.activation.message"
```

**Comment by heimrih** [comment]
ready

**Review by saobv-2223 — COMMENTED** [review]
merge

**Pull request merged**